### PR TITLE
add docs for evaluating with attachments

### PR DESCRIPTION
**Pull request by isahers1**

### CODE REVIEW

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+# Define the example
+example = ExampleUpsertWithAttachments(
+    dataset_id=dataset.id,
+    inputs={"question": "What were the cumulative earnings earned from online orders in the midwest during Q2?"},
+    outputs={"answer": "$123456"},
+    attachments={
+        # Each attachments is just a name with a mime type and the bytes content of the file
+        "pdf": ("application/pdf", Path("./foo_earnings.pdf").read_bytes()),
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:37]: not related to this pr, but is it too late to change tuple -> dict? imo naming the values improves usability a lot, eg
```python
attachments = [
  {
    "name": "my_pdf",
    "mime_type": "applications/pdf",
    "bytes": Path(...).read_bytes()
  }, 
  ...
]
```
is easier to read for me

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+Attachments allow you to associate large files with your examples. This allows you to evaluate RAG applications
+over large internal documents, benchmark image analysis tools, etc.
+
+## Create a dataset with attachments
+
+To create a dataset with attachments, you need to use the `upsert_examples_multipart` method of the LangSmith client:
+
+```python
+from langsmith.client import Client
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:19]: can import Client directly form langsmith

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import {
+  CodeTabs,
+  python,
+  typescript,
+  PythonBlock,
+  TypeScriptBlock,
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:6]: uneeded

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+} from "@site/src/components/InstructionsWithCode";
+
+# Evaluate an LLM application with attachments
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:9]: feels like this describes the implementation not the problem. maybe `Evaluate applications on large files` or `Evaluate applications with large file inputs` or something like that

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+Attachments allow you to associate large files with your examples. This allows you to evaluate RAG applications
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:11]: maybe worth adding a sentence about why attachments are better than storing file in example inputs

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    attachments={
+        # Each attachments is just a name with a mime type and the bytes content of the file
+        "pdf": ("application/pdf", Path("./foo_earnings.pdf").read_bytes()),
+        # We can pass multiple attachments (of different types!), as long as they have different names
+        "pptx": ("application/pptx", Path("./foo_earnings.pptx").read_bytes()),
+    },
+)
+
+# Upsert the examples
+langchain_client.upsert_examples_multipart(upserts=[example])
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:44]: ooc could this just be part of langchain_client.create_example? and we route to multipart upsert if attachments are passed in?

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        img_bytes = BytesIO()
+        pix.pil_save(img_bytes, format=image_format)
+        # Encode the bytes in base64
+        base64_bytes = base64.b64encode(img_bytes.getvalue()).decode('utf-8')
+        images.append(base64_bytes)
+    pdf_document.close()
+    return images
+
+
+def target(inputs, attachments):
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:72]: nit: would give this a name that indicates what it is (eg `pdf_qa`)

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    return {"answer": model.invoke(messages).content}
+```
+
+## Run an evaluation
+
+We can then run an evaluation as usual, by passing in the target function to the `evaluate` method:
+
+```python
+# We can optionally define an evaluator to use
+def evaluator(run, example):
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:94]: can we use the new evaluator format eg (`evaluator(outputs: dict, reference_outputs: dict)`)

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+The target function must have two positional arguments, the first must be called `inputs` and the second must be called `attachments`.
+
+```python
+from langchain_openai import ChatOpenAI
+from langchain_core.messages import HumanMessage
+
+model = ChatOpenAI(model="gpt-4o-mini")
+
+def pdf_to_image_bytes(pdf_bytes, image_format='PNG'):
+    pdf_document = fitz.open(stream=pdf_bytes, filetype="pdf")
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:59]: need to import fitz

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+## Create a dataset with attachments
+
+To create a dataset with attachments, you need to use the `upsert_examples_multipart` method of the LangSmith client:
+
+```python
+from langsmith.client import Client
+from langsmith.schemas import ExampleUpsertWithAttachments
+
+# Pass in your api key directly, or define it in the LANGCHAIN_API_KEY environment variable
+langchain_client = Client(api_key="...")
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:23]: nit: would call it ls_client or langsmith_client

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx (hunk 6 above)
Comment (isahers1) [reply docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:44]: not sure @agola11 might have thoughts here

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx (hunk 1 above)
Comment (baskaryan) [reply docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:36]: related to this pr, would be nice to download some real public pdf here to use. could just be https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+## Define a target function with attachments
+
+Now that we have a dataset that includes examples with attachments, we can define a target function to run our LLM application with the attachments.
+The target function must have two positional arguments, the first must be called `inputs` and the second must be called `attachments`.
+
+```python
+from langchain_openai import ChatOpenAI
+from langchain_core.messages import HumanMessage
```
Comment (baskaryan) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx]: import SystemMessage

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+This also avoid potential performance bottlenecks with parsing large JSON payloads.
+
+Finally, attachments are more user-friendly in the LangSmith UI, as they are rendered as files with previews, rather than as base64-encoded strings.
+:::
+
+## Create a dataset with attachments
+
+### SDK
+
+To create a dataset with attachments using the SDK, you need to use the `upsert_examples_multipart` method of the LangSmith client.
```
Comment (agola11) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:20]:
```suggestion
To create a dataset with attachments using the SDK, you need to use the `upload_examples_multipart` method of the LangSmith client.
```

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+# Upsert the examples
+langsmith_client.upload_examples_multipart(uploads=[example])
+```
+
+Once you upload examples with attachments, you can view them in the LangSmith UI. Each attachment will be rendered as a file with a preview, making it easy to inspect the contents.
+![](./static/attachments_with_examples.png)
+
+### From existing runs
+
+You can also add examples with attachments from existing runs that contain attachments.
```
Comment (agola11) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:70]:
```suggestion
When adding runs to a LangSmith dataset, attachments can be selectively propagated from the source run to the destination example.
```

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+langsmith_client.upload_examples_multipart(uploads=[example])
+```
+
+Once you upload examples with attachments, you can view them in the LangSmith UI. Each attachment will be rendered as a file with a preview, making it easy to inspect the contents.
+![](./static/attachments_with_examples.png)
+
+### From existing runs
+
+When adding runs to a LangSmith dataset, attachments can be selectively propagated from the source run to the destination example.
+To do learn how to do so you can read [this guide](./../datasets/manage_datasets_in_application#add-runs-from-the-tracing-project-ui).
```
Comment (agola11) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:71]:
```suggestion
To do learn more, please see [this guide](./../datasets/manage_datasets_in_application#add-runs-from-the-tracing-project-ui).
```

File: docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+### From the LangSmith UI
+
+You can also add examples with attachments from the LangSmith UI. You can do so by clicking the `+ Example` button in the `Examples` tab of the dataset UI.
+You can then upload the attachments that you want by using the "Upload Files" button:
+
+![](./static/create_example_with_attachments.png)
+
+## Running evaluations with attachments
+
+Once you have a dataset with attachments, you can run evaluations with attachments as follows:
```
Comment (agola11) [inline docs/evaluation/how_to_guides/evaluation/evaluate_with_attachments.mdx:82]:
```suggestion
Once you have a dataset that contains examples with file attachments, you can run evaluations that process these attachments.
```

**Pull request merged**